### PR TITLE
Discussion: record-elektra use cases

### PR DESCRIPTION
**Pull request by atmaxinger**
This PR should contain the major use cases/features for the following scenario:

- Export existing configuration into an Ansible playbook to be applied to other machines also running Elektra
- Record changes on any machine configured using above playbook
- Export those changes to yet another playbook
- Be able to apply those changes to all the other configured machines

To focus on the overall picture of this scenario I have currently intentionally left out all the different use cases we have already talked about for the record-elektra tooling itself - like session handling. See #4399.

### CODE REVIEW

File: doc/usecases/record_elektra/UC_change_notifications.md
Comment (kodebach) [inline doc/usecases/record_elektra/UC_change_notifications.md:9]: IMO this is to general. To me this reads like: "A plugin whose main purpose is something else (e.g. `type`) can be notified of changes via an extra API." I think that would lead to chaos.

IMO what we should have is:

> Some plugins are only executed when something changed in the KDB (*) and as an input receive the keys that changed in their old and new state.

Importantly, the plugins should already receive the changeset as an input. In other words, `libelektra-kdb` (or an extra component between `libelektra-kdb` and the plugins) does the change tracking not the plugins themselves. This is important, because otherwise every such "change plugin" would keep its own copy of all keys to keep track of the old values.

---

(*) This is already the case for standard mounted backend plugins. They are only called, if there is a changed key in their mountpoint. But they always receive all keys for the mountpoint and don't receive the old state.

**Review by kodebach — COMMENTED** [review]
Looks good to me. I have a small note on what actually is recorded. I'd also like some clarification on the "notify plugins" thing.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:5]: Useless to state the same title again.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:8]:
```suggestion
- **Actors:** System administrator
```

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:9]: It is still not clear to me why we the system administrator ansible-libelektra to fail? What is the goal? Give a bit background why the user needs this. Give examples.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:13]:
```suggestion
- **Precondition:** The user marks a certain key in the task
```

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:17]: Isn't it a postcondition that we only continue if the value is as wanted?

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md:9]: Not only to specify but that ansible-libelektra actually does the merging

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md:16]: Describe what can go wrong in merging...

File: doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md:8]: typo

File: doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md:1]:
```suggestion
# Use Case: Remove keys in ansible-libelektra
```

File: doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_remove_keys.md:17]: :+1:

File: doc/usecases/record_elektra/UC_ansible-libelektra_report_conflicts.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_report_conflicts.md:1]: One merge use case should be enough? This is the Error scenario of the merge use case?

File: doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md:17]:
```suggestion
- **Postcondition:** All locally changed keys get tracked.
```

File: doc/usecases/record_elektra/UC_ansible_export.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible_export.md:13]: Isn't there some valid "empty" playbook?

File: doc/usecases/record_elektra/UC_ansible_export.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible_export.md:23]: This UC does not catch the situation that a playbook is updated (already existed before and gets updated with what was recorded).

File: doc/usecases/record_elektra/UC_ansible_specification.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible_specification.md:16]: Is this export also connected to UC_ansible_export.md?

File: doc/usecases/record_elektra/UC_ansible_specification.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible_specification.md:15]: Not clear to me.

File: doc/usecases/record_elektra/UC_ansible_specification.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible_specification.md:20]: What happens?

File: doc/usecases/record_elektra/UC_cmerge_conflict_keys.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_cmerge_conflict_keys.md:1]: Probably too low-level to be needed as UC. It is simply missing functionality, needed for the merge UC.

File: doc/usecases/record_elektra/UC_change_notifications.md
Comment (markus2330) [reply doc/usecases/record_elektra/UC_change_notifications.md:9]: I agree, we need this narrowed down, maybe even several UC (logging, change notification, ...)

**Review by markus2330 — COMMENTED** [review]
Thank you for the PR! I am missing the high-level use cases (build server, nextcloud, olimex should be somehow derivable from the UC described here). What is the workflow you plan to do while doing the three UC? At least the "do local changes/updating playbook" UC is missing.

File: doc/usecases/record_elektra/UC_change_notifications.md
Comment (kodebach) [reply doc/usecases/record_elektra/UC_change_notifications.md:9]: > logging, change notification

I don't think that separation would make sense, since "logging" is an application/special case of "change notification". The plugin is notified of changes and just logs them somewhere, instead of doing something else (e.g. broadcasting via dbus)

However, it might make sense to make one UC for "global/hook plugins" in general and a second UC for a "logging plugin". But this may be a too technical distinction, as "global/hook plugins" could just be considered the implementation details of the current UC.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:9]: This can be used as a safe guard to ensure the configuration is only applied to the correct machines. Probably won't be used by many, but it could be useful to some. This the feature needed for `kdb record mark` to make any sense.

Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:9]: A UC should clearly state for whom, in which situation and why it is useful.

Obviously we cannot know how often features will actually be used. For now there should be at least one convincing example where we ourselves would use it (or at least can imagine using it).

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:5]: But the template clearly states so...

grafik

File: doc/usecases/record_elektra/UC_ansible-libelektra_report_conflicts.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_report_conflicts.md:1]: Depends on how granular these use cases should be. But yes, we can move them into the same use case.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (kodebach) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:9]: I think one example would be a dependency between two applications. Let's say apps A and B work together. There could a key in A's config that may be set to tell A to work with B. Now we define a config for B and this config only works, when used with A. Therefore, we need A to configured to work with B.

We could of course just set A's key to the right value and include that in the recording. But IMO that would really be good practice. We'd be reconfiguring A when we apply the config for B. Instead a better solution would be just marking the the key in A's config and thereby recording that it must be set correctly for this recording to be applicable.

---

Another example that isn't covered by this UC, but which might be useful (and is probably more common): Let's say there is a key that can't be set automatically, but must be manually configured by the sysadmin before applying the config. This could for example be a path that is unique to the machine. In this case, we'd need away to record "this key must be set, but we don't care about the value".

However, this special case can probably be better achieved, by adding an additional task to the Ansible playbook.

---

The above are the most reasonable examples I could think of, but there are other more questionable cases. For example, if setting the key in question is not idempotent, e.g. because of some (less than ideal) plugin used for the mountpoint.

---

> **Note** On a technical level, the ability to record such "marked keys" falls directly out of the "record old & new values" format. For a marked key, the old and new value would just be identical. Therefore, what we are really deciding on with this use case is whether or not to expose this feature to the user.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:5]: As you want :wink:

File: doc/usecases/record_elektra/UC_ansible_export.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible_export.md:13]: Yes! we could just create that.

File: doc/usecases/record_elektra/UC_ansible_export.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible_export.md:23]: Wouldn't this be another use case? Merging of playbooks?

File: doc/usecases/record_elektra/UC_ansible_specification.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible_specification.md:16]: Yes! The idea of this use case was/is that we could (sometimes in the future) generate "real" Ansible playbooks that work without ansible-libelektra. But I think this is something out of scope for now.

**Comment by atmaxinger** [comment]
@markus2330 I have now created a new use case for the whole "do local changes/updating playbook" --> `UC_interactive_configuration.md`

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_interactive_configuration.md:7]: This must be somewhere else, otherwise "Level:" is useless.

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_interactive_configuration.md:8]: Be as specific as possible. Who do we target?

```suggestion
- **Actors:** System administrator
```

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_interactive_configuration.md:10]: Why limit to modifications on a single host? As long the changes can somehow be merged...

```suggestion
It should be possible to modify the configuration directly on hosts, and then apply the changes onto all other hosts.
```

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_interactive_configuration.md:21]:
```suggestion
- **Alternative scenario:** If the host has no applied playbook, it is only possible to export a playbook with the changed keys
```

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md:9]: Yes, I agree: limited/broken plugins, in particular read-only plugins, are a situation I could imagine it makes sense. There are some cases where plugins must be read-only by design (hardware-information, read-only file systems etc.) so in these cases assertions are the only way.

Btw. terminology-wise we used "assert" in #4399. This alone would make the UC much clearer.

**Review by markus2330 — COMMENTED** [review]
This use case was very much needed, good work! What I am still missing is a UC that somehow would make multiple sessions necessary. From the current UCs actually a single global session would be enough.

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_interactive_configuration.md:10]: The problem is, if you change the configuration on multiple systems at the same time, we'd have to manually merge the playbooks - as in the user would have to copy all the playbooks from the machines and let a tool merge them. 

If we limit this to a single host at a time, the ansible export plugin could do this automatically at this host.

File: doc/usecases/record_elektra/UC_interactive_configuration.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_interactive_configuration.md:21]: Hmm we can't do it with every playbook. It must be a playbook that was specifically exported by our plugin to do this, as we need to copy the playbook to the remote hosts for the merging to work.

Comment (markus2330) [reply doc/usecases/record_elektra/UC_interactive_configuration.md:21]: I did not intend to change the meaning in this way. You can simply ignore the suggestion.

File: doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_marked_keys.md]: Another example would be to check if a certain Elektra plugin is installed, e.g. the `check` plugin.

File: doc/usecases/record_elektra/UC_change_notifications.md
Comment (kodebach) [reply doc/usecases/record_elektra/UC_change_notifications.md]: >  log read-only accesses to the kdb, maybe for audit purposes

Yeah, I think this would be very far out of scope. Logging read accesses a totally different can of worms.

What should be possible with the basic logging/notification hook is logging the parent key that was used in every `kdbGet`. Any logging/notification plugin would be called with the parent key anyway.

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md:15]: This is vague. It would be better if you decide.

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md:9]: Please list at least the currently known strategies we want to implement. Ideally the use case gives argumentation when to use what.

File: doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md:9]:
```suggestion
- **Brief:** After a new elektra configuration has been rolled out via Ansible, automatically enable or reset session recording for the appropriate namespaces
```

File: doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md:14]: Sometimes there are dots, sometimes not?

```suggestion
- **Main success scenario:** User deploys configuration via Ansible
```

File: doc/usecases/record_elektra/UC_change_notifications.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_change_notifications.md:18]: Isn't this implementation detail? Actually they want to know the difference? (Duplicate with Brief)

File: doc/usecases/record_elektra/UC_record_changes.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_record_changes.md:18]:
```suggestion
  - The old & new values for every changed key and metakey are recorded.
```

File: doc/usecases/record_elektra/UC_record_changes.md
Comment (markus2330) [inline doc/usecases/record_elektra/UC_record_changes.md:20]: I think we should make `kdb set` warn about this.

File: doc/usecases/record_elektra/README.md
Comment (markus2330) [inline doc/usecases/record_elektra/README.md:7]: move this section down

**Review by markus2330 — CHANGES_REQUESTED** [review]
It looks much better now! The PR is nevertheless still marked as draft. Please undraft and set "ready to merge" so that we can get this merged eventually.

One UC I would find very interesting is to allow specific overrides https://github.com/ElektraInitiative/ansible-libelektra/issues/11 and nevertheless allow exports (i.e. we would need some way to tell if a change is specific to a host or should be applied to all hosts).

File: doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md:9]: What would 'the appropriate namespaces' be?

File: doc/usecases/record_elektra/UC_record_changes.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_record_changes.md:20]: Yes, my bad. This is what I meant by 'We log'.

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md:15]: It think we should use OURS as default. It's probably the intention in 99% of the use cases to set a new value when using Ansible.

Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: I agree!

File: doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md
Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_start_recording.md:9]: Probably all of them which had changes (any recording happened)?

File: doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md
Comment (kodebach) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: I'm not sure, using "ours" to resolve conflicts by default is a good idea. IMO the default should be to just abort. Otherwise things may already be broken before the user even notices that there was a merge.

Yes, most people will probably decide to use "ours", but it's not a safe default IMO. It's kind of as if the default for an SQL INSERT would be to replace a conflicting primary key instead of aborting.

Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: Yes, abort on conflicts is probably a better choice. Then people look at which changes were done by maintainers and manually select what is actually best.

Btw. we discarded the undo UC. It might help with unwanted merges but it is probably not worth the effort to keep previous recordings.

Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: I'm not even sure anymore if we need the merging for the Ansible Use Case. If I execute an Anisble playbook, my goal is that all the hosts I run them on are in the configured state. After all, an Ansible playbook is supposed to be a single source of truth.

Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: I mean I'll still implement this feature ... but for the Case Studies this shouldn't be needed.
Also, to be usable we need to supply the `base`. Otherwise, it will always come to conflicts (except keys that never change) if we use `ABORT` as default.

Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: As I envisioned it, Ansible runs would set a new BASE. Modifications of Ansible playbooks are OURS, modifications on the machine are THEIRS.

THEIRS can have two reasons:
1. manual interventions as supported by our recording approach
2. updates of software (i.e. new config files in /etc installed by packages)

Because of case 2. this UC is very much needed whenever you will use ansible-libelektra (regardless of if you also use record-libelektra). If 1. is done correctly (all changes exported and added to playbooks) there actually should be no conflict, as OURS would be THEIRS (both different from BASE). If someone forgot exporting and the playbook introduces incompatible changes, the strategy suggested from @kodebach, would loudly communicate that the sys-admins did something wrong, which looks like to be the proper way to do it. (Careful sys-admins would reset the machines to the previous BASE if they played around and nothing useful came out.)

Did this clarify the concerns?

Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: Yes this did clarify a few concerns. I totally forgot about the updates of software.

The only thing I still don't get is what you mean by 'Ansible runs would set a new BASE'. Where is this BASE stored? Within the playbook?

Comment (markus2330) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: It cannot be the playbook (otherwise you would not be able to reset to base locally) but where exactly is actually implementation detail. I thought that when `kdb record` gets reset, "/elektra/record" (or whatever we use as path for recording) is empty and with every changed key, the old values/metadata of that key get stored in "/elektra/record". So basically the merge from the current keys and (with higher priority) /elektra/record is the base. But let us discuss details later when you actually come to the implementation. For now hooks/global plugins have highest priority :rocket:

Comment (atmaxinger) [reply doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md]: This is actually a good idea - but remember: We don't track mounted configuration files. So if an update changes those directly we are currently out of luck.

I agree, let's get this merged and concentrate on the hooks.

**Review by markus2330 — APPROVED** [review]
Looks much better. Some parts need to be revisited later (which merge strategies, details of change notifications) but in general the UC goes to a good direction.

@atmaxinger For this PR: please have a look at the build server problems.

**Comment by atmaxinger** [comment]
Jenkins build libelektra please

**Comment by atmaxinger** [comment]
@markus2330 the BSD builds fail because of the wrong version of `libc`, and the jenkins build is flakey. For example, right now its running into a timeout when importig demo.kdb. Not much I can do about that in this PR.

**Comment by markus2330** [comment]
Please report problems of Jenkins to #160.

@Eiskasten can you help out with the BSD failures?

**Comment by markus2330** [comment]
@atmaxinger please report problems of Jenkins to https://github.com/ElektraInitiative/libelektra/issues/160.

@Eiskasten can you help out with the BSD failures?

**Comment by atmaxinger** [comment]
jenkings build libelektra please

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
Yes, let us merge it. Great work! :fireworks:

**Pull request merged**
